### ISSUE TITLE
Product quantity with units

### ISSUE THREAD

**Issue by SeSamSeg** [PrestaShop/PrestaShop#12334]
**I'd like to add a product with a quantity in meters instead of "Units" or a multiple of pieces instead of 1**

e.g.:

- Fabric in multiples of 0.5m.
- By weight per 100g, in kg in the store
- Boxes of 25

**Add unit multiplier, unit steps and unit name**

I put together a way to do this in my shop with:

Multiplier: (the amount to multiply with for the frontend)

- Fabric can be expressed in cm in the backend, and with a multiplier of 0.01 for the frontend.
- Weight items can be expressed in g in the backend, and with a multiplier of 0.001 for the frontend.
- Boxes can be put in the backend as boxes of 25 and with a multiplier of 25 in the frontend so they get 100pcs instead of 4x box of 25pcs.

Unit step: (the amount to step with the + and - in the frontend and as a check in the backend)

- Fabric per 0.5m = 50cm so a step of 50.
- Weight per 0.1kg = 100g so a step of 100.
- Per 1 Box of 25 pcs so a step of 1.

Unit name for front and one for backend: (should be defferent per language and maybe dependant on the amount?)

- Fabric: backend: cm, frontend m
- Weight: backend: g, frontend kg
- Boxes: Backend: 'boxes of 25', frontend: 'pcs'

**An alternative is using float instead of multipliers**

The quantity in the db can also be expressed as floats, but then I still miss the stepping and units. There is also a higher chance of precision loss with floats and if you put the quantity with a good multiplier, you always have a number. 

**I have this currently in use**

I have a hacky way of doing it now in the backend. I tried to do it only with a addin, but this is unfeasible because of rounding errors in the backend code.

I'm going to clean it up the code a bit in my shop, if there is a chance that a pull request for this could be merged, I'd spend a bit more time on it to get it in shape for merging. If not, I'd just add it as an override and do a lot of things manually.

**Comment by khouloudbelguith** [comment]
Hi @labsin,

Thanks for your report.
@marionf, @colinegin what do you think of his suggestion?
Thanks!

Related: #11632

**Comment by colinegin** [comment]
Hello,

Thanks for your idea ! Would you be able to do a PR for this improvement ? 
Our dev/UX/product team will review it asap once the commit is done.

Thanks !

**Comment by MatShir** [comment]
Hi @labsi, 

You might not have the technical skills to provide it to Prestashop. But we got your feedback and we have shared the idea at Prestashop Feedbacks (http://feedback.prestashop.com/) to let the community help us to prioritizes all ideas from the community.

Thank you

**Comment by kpodemski** [comment]
Hey @colinegin 

i don't think PrestaShop will ever decide to accept such a modification because we would need to change database structure and all quantities in INTEGER would become FLOAT

if @eternoendless will give it green light i'll be happy to do such a modification because i already have one working, tho this is a big one, core, BO, front-office, JS...

**Comment by colinegin** [comment]
@eternoendless @PrestaShop/prestashop-core-developers what do you think ?

**Comment by SeSamSeg** [comment]
Hi

I have this implemented in my shop (not yet published)

I've added a pull request with changes based on 1.7.4.1:
https://github.com/PrestaShop/PrestaShop/pull/12665

![schermafdruk_2019-02-21_22-37-24](https://user-images.githubusercontent.com/1176649/53203400-5d657500-3629-11e9-9784-94c929509e9d.png)
![schermafdruk_2019-02-21_22-34-08](https://user-images.githubusercontent.com/1176649/53203401-5d657500-3629-11e9-9bb1-b135a7e2a775.png)

**Comment by JjPrus** [comment]
Hi man!
Can you manage to achive working version? 
I have exacly the same problem and looking for some help  :)

**Comment by khouloudbelguith** [comment]
Hi @JjPrus,

This issue not fixed yet.

Thanks!

**Comment by matks** [comment]
> i don't think PrestaShop will ever decide to accept such a modification because we would need to change database structure and all quantities in INTEGER would become FLOAT


Hi @kpodemski when you said that you were not a PrestaShop maintainer. Now you are one 😄  so... what do you think?

As for me, I have an opinion that I think everybody will dislike 😄 : this issue has been opened 2 years ago, and nothing happened. PrestaShop did not use units for years and few people complained. So I understand that _some people_ would be happy to have this feature inside PrestaShop but I am not convinced this is a must-have, awaited by hundred of users.

On the other hand, _enabling this feature requires quite a heavy modification of the database_ as we can look in PR https://github.com/PrestaShop/PrestaShop/pull/12665/ submitted by @labsin . This means a quite a lot of work to setup, upgrade and then maintain all of this.

So
- on one hand we have a feature that - as far as I know - is not needed by a lot of people
- on the other hand it requires a database structure modification that will impact everybody

I'm afraid that _in my opinion this is not worth it_. So I think it should not be integrated in the Core. Obviously PrestaShop developers are free to implement it on their shop 😉  but I dont feel like this is something that will benefit the 300,000 shops powered by PrestaShop.

This is why I close this issue.

@kpodemski I'm open to discuss and change my mind if you want to talk 😊

**Comment by SeSamSeg** [comment]
> On the other hand, _enabling this feature requires quite a heavy modification of the database_ as we can look in PR #12665 submitted by @labsin . This means a quite a lot of work to setup, upgrade and then maintain all of this.

I think a separate linked table for units only for the products when needed is enough. The PR used multipliers, just the value that 1 unit represents (multiplier of 0.01 means 1 unit is 0.01m) and a unit name. This could even be implemented as a plugin and isn't a lot of work and doesn't need a float quantities.

I also added a minimum quantity en steps, like you want 300 bolts and they are minimum order of 100 in steps of 50. This needs changes in the core.

> I'm afraid that _in my opinion this is not worth it_. So I think it should not be integrated in the Core. Obviously PrestaShop developers are free to implement it on their shop 😉 but I dont feel like this is something that will benefit the 300,000 shops powered by PrestaShop.

I think most that want units and minimum quantities and steps, just aren't going to use prestashop when they see it it's supported. In my shop I've had 3 different kind of products where I wanted to use these features (boxes, weighted products and fabric) so I don't tink it's too far fetched. I've seen online a couple of people trying to change Qty to float, so there is some animo around the idea.

I implemented it as a POC and then moved on with other thinks because I did't tink it'd me implemented and I found the quantity handling in core a bit difficult and I wouldn't have kept it up to date and barely got it to work.
